core/sql/optimizer/BindRelExpr.cpp

```diff
@@ -10171,17 +10164,26 @@ matched clause of merge). If the upsert caused a row to be updated in the
 base table then the old version of the row will have to be deleted from 
 indexes, and a new version inserted. Upsert is being transformed to merge
 so that we can delete the old version of an updated row from the index.
+
+Upsert is also converted into merge when there are omitted cols with default values and 
+TRAF_UPSERT_WITH_INSERT_DEFAULT_SEMANTICS is set to  OFF in case of aligned format table or 
+omitted current timestamp cols in case of non-aligned row format
```

Comment:
For upsert, in case of non-aligned format, if any column other than the columns with default current is omitted, the default value is resolved at the time of select. If the current default column is omitted it will be always changed to merge. User can't force a different behavior for non-aligned format in this specific case. Rationale here: If we allow user to control it, then all omitted default columns should always be populated at the time of upsert for the merge to work correctly.

For non-aligned format, user control it using the above cqd. Default behavior is change it to merge if the default columns are omitted in upsert.